### PR TITLE
WIP: Add Typings for the public SDK of Scrivito CMS

### PR DESCRIPTION
**Pull request by juliankrieger**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [x] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [ ] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

If removing a declaration:
- [ ] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

**Comment by juliankrieger** [comment]
Alright, seems like all checks are passed.

**Comment by juliankrieger** [comment]
Closing PR temporarily to clarify if this is GPL3 compliant.

**Review by krishan — APPROVED** [review]
Hi, Scrivito developer here. We are hoping to release official types at some point in the future (sorry we did not yet get around to it), but for now this looks great 👍

File: types/scrivito/lib/Attribute.d.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
+export type Attribute =
```
Comment (DanielRosenwasser) [inline types/scrivito/lib/Attribute.d.ts:1]: These `lib` files don't actually exist in the package. Can you inline them please?

File: types/scrivito/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
+// Type definitions for Scrivito 1.16
```
Comment (DanielRosenwasser) [inline types/scrivito/index.d.ts:1]: Can you lowercase the package name?

@sandersn should the header parser be stricter about this?

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]

**Review by krishan — APPROVED** [review]

File: types/scrivito/lib/Attribute.d.ts (hunk 1 above)
Comment (juliankrieger) [reply types/scrivito/lib/Attribute.d.ts:1]: yes, I'll see to it today. What do you think about a type only export (provided we use Typescript 3.8 and newer)

File: types/scrivito/index.d.ts (hunk 2 above)
Comment (juliankrieger) [reply types/scrivito/index.d.ts:1]: I'll do it anyway for correctness

File: types/scrivito/lib/Attribute.d.ts (hunk 1 above)
Comment (juliankrieger) [reply types/scrivito/lib/Attribute.d.ts:1]: @DanielRosenwasser I checked, the typings you see are exported only in Attribute.d.ts to be used in other definition files. They are not exported to the public facing API, e.g you can not do

```
import { Attribute } from 'scrivito'
```

**Comment by juliankrieger** [comment]
@DanielRosenwasser I implemented all your change requests. If you have any more please let me know

**Comment by juliankrieger** [comment]
@krishan can we talk about how to type the `propertiesGroups` parameter in `provideEditingConfig`? As I see it, there are two scenarios:
- `properties` if nothing else but `title` is given
- `component`: this makes the `properties` paramter optional.

We could also think about adding type only exports, to make development easier if you need to calculate function parameters before passing them. If we exported the types of those parameters, typing these pre-calculated objects would become easier.

export type {
    AttributeOptions,
    AttributeProps,
    AttributeValue,
    BackgroundImageBackgroundProp,
    BackgroundImageTagProps,
    ChildListTagProps,
    ConfigOptions,
    ContentTagProps,
    CreateAttributes,
    CSSImageStyleBackgroundProps,
    EditingConfig,
    EditingConfigAttributes,
    ExternalLink,
    ExternalLinkAttributes,
    ImageTagProps,
    InternalLink,
    InternalLinkAttributes,
    LinkTagProps,
    ObjClassOptions,
    ObjComponentProps,
    OptimizeDefinition,
    PropertiesGroup,
    WidgetClassOptions,
    WidgetComponentProps,
    WidgetTagProps
}

**Comment by elibarzilay** [comment]
@juliankrieger: is this ready or is it's still a WIP?  (If it is, then it's better to keep convert to a draft pr)

**Comment by orta** [comment]
( We've switched DT rotation, so please remember to give us a heads up when it's good to review )

**Comment by juliankrieger** [comment]
@orta With scrivito developers not having any objections, I think we're ready to go!

**Comment by orta** [comment]
👍🏻

**Pull request merged**